adal/src/main/java/com/microsoft/aad/adal/AuthenticationException.java

```diff
@@ -70,6 +84,89 @@ public AuthenticationException(ADALError code, String details) {
     public AuthenticationException(ADALError code, String details, Throwable throwable) {
         super(details, throwable);
         mCode = code;
+
+        if (null == throwable) {
+            return;
+        }
+
+        if (throwable instanceof AuthenticationException) {
+            mServiceStatusCode = ((AuthenticationException) throwable).getServiceStatusCode();
+            mResponseBody = ((AuthenticationException) throwable).getResponseBody();
+            mHttpResponseHeaders = ((AuthenticationException) throwable).getHttpResponseHeaders();
+        }
+    }
+
+    /**
+     * @param code Resource file related error code. Message will be derived
+     *            from resource with using app context
+     * @param details Details related to the error such as query string, request info.
+     * @param response HTTP web response
+     */
+    public AuthenticationException(ADALError code, String details, HttpWebResponse response) {
+        super(details);
+        mCode = code;
+
+        if (null == response) {
+            return;
+        }
+
+        mServiceStatusCode = response.getStatusCode();
+
+        if (null != response.getResponseHeaders()) {
+            mHttpResponseHeaders = response.getResponseHeaders();
+        }
+
+        if (null != response.getBody()) {
+            final String responseBodyStr = response.getBody();
+            try {
+                mResponseBody = new HashMap<>();
+                final JSONObject jsonObject = new JSONObject(responseBodyStr);
+                final Iterator<?> i = jsonObject.keys();
+                while (i.hasNext()) {
+                    final String key = (String) i.next();
+                    mResponseBody.put(key, jsonObject.getString(key));
+                }
+            } catch (final JSONException exc) {
+                return;
+            }
+        }
+    }
+
+    /**
+     * @param code Resource file related error code. Message will be derived
+     *            from resource with using app context
+     * @param details Details related to the error such as query string, request info.
+     * @param response HTTP web response
+     * @param throwable {@link Throwable}
+     */
+    public AuthenticationException(ADALError code, String details, HttpWebResponse response,
+                                   Throwable throwable) {
+        this(code, details, throwable);
+
+        if (null == response) {
+            return;
+        }
+
+        mServiceStatusCode = response.getStatusCode();
+
+        if (null != response.getResponseHeaders()) {
```

Comment:
don't duplicate code

Reply:
Sorry, do you mean Line 148 is duplicated? Addressed.

Reply:
No, this function and the previous one have the same code repeated. Line 111-122 and 141 to 152 are same